src/csrf.ts

```diff
@@ -0,0 +1,204 @@
+    this.session = session
+    this.options = options
+    this.applicationKey = applicationKey
+  }
+
+  /**
+   * Get the request method, check if the user defined
+   * methods allowed for csrf verification in the
+   * config. If it did, check if the request
+   * method is one of the allowed. If not,
+   * return false.
+   */
+  private requestMethodShouldEnforceCsrf (request: RequestContract): boolean {
+    const method = request.method().toLowerCase()
+
+    if (!this.options.methods || this.options.methods.length === 0) {
+      return true
+    }
+
+    return this.options.methods
+      .filter(definedMethod => definedMethod.toLowerCase() === method)
+      .length > 0
+  }
+
+  /**
+   * Check if the current request url has been
+   * excluded from csrf protection.
+   */
+  private requestUrlShouldEnforceCsrf (ctx: HttpContextContract): boolean {
+    if (!this.options.exceptRoutes || this.options.exceptRoutes.length === 0) {
+      return true
+    }
+
+    return !this.options.exceptRoutes.includes(ctx.route!.pattern)
+  }
+
+  /**
+   * Check if csrf secret has been saved to
+   * session. If not, generate a new one,
+   * save it to session, and return it.
+   */
+  public async getCsrfSecret (): Promise<string> {
+    let csrfSecret = this.session.get('csrf-secret')
+
+    if (!csrfSecret) {
+      csrfSecret = await Csrf.secret()
+
+      this.session.put('csrf-secret', csrfSecret)
+    }
+
+    return csrfSecret
+  }
+
+  /**
+   * Extract the csrf token from the request by
+   * checking headers and inputs. Decode the
+   * token if it was encrypted.
+   */
+  private getCsrfTokenFromRequest (request: RequestContract): string|null {
+    const token = request.input('_csrf') || request.header('x-csrf-token')
+
+    if (token) {
+      return token
+    }
+
+    const encryptedToken = request.header('x-xsrf-token')
+    const unpackedToken = encryptedToken ? unpack(token, this.applicationKey) : null
+
+    return unpackedToken ? unpackedToken.value : null
+  }
+
+  /**
+   * Generate a new csrf token using
+   * the csrf secret extracted
+   * from session.
+   */
+  public generateCsrfToken (csrfSecret): string {
+    return Csrf.create(csrfSecret)
+  }
+
+  /**
+   * Set the xsrf cookie on
+   * response
+   */
+  private setXsrfCookie (ctx: HttpContextContract): void {
+    ctx.response.cookie('x-xsrf-token', ctx.request.csrfToken)
+  }
+
+  /**
+   * Set the csrf token on
+   * request
+   */
+  private setCsrfToken (ctx: HttpContextContract, csrfSecret: string): void {
+    ctx.request.csrfToken = this.generateCsrfToken(csrfSecret)
+  }
+
+  /**
+   * This would make a csrfToken variable available to
+   * the edge view templates. This would also create
+   * a helpful method called csrfField to be used
+   * on the frontend to generate a hidden
+   * field called _csrf
+   */
+  private shareCsrfViewLocals (ctx: HttpContextContract): void {
+    ctx.view.share({
```

Comment:
Just add a conditional here to check that `ctx.view` is defined. In some rare edge cases, it is possible that someone is used CSRF but without the views, or maybe they swapped the template engine to something else